File: src/main/java/org/taktik/icure/dao/MessageDAO.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao;

import org.ektorp.support.View;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.Message;

import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public interface MessageDAO extends GenericDAO<Message> {

	List<Message> findByFromAddressActor(String partyId, String fromAddress, List<String> actorKeys);

	List<Message> findByToAddressActor(String partyId, String toAddress, List<String> actorKeys);

	List<Message> findByTransportGuidActor(String partyId, String transportGuid, List<String> actorKeys);

	PaginatedList<Message> findByFromAddress(String partyId, String fromAddress, PaginationOffset<List<Object>> paginationOffset);

	PaginatedList<Message> findByToAddress(String partyId, String toAddress, PaginationOffset<List<Object>> paginationOffset, Boolean reverse);

	PaginatedList<Message> findByHcParty(String healthcarePartyId, PaginationOffset<List<Object>> paginationOffset);

	PaginatedList<Message> findByTransportGuid(String healthcarePartyId, String transportGuid, PaginationOffset<List<Object>> paginationOffset);

	PaginatedList<Message> findByTransportGuidReceived(String healthcarePartyId, String transportGuid, PaginationOffset<List<Object>> paginationOffset);

	PaginatedList<Message> findByTransportGuidSentDate(String healthcarePartyId, String transportGuid, Long fromDate, Long toDate, PaginationOffset<List<Object>> paginationOffset);

	List<Message> findByHcPartyPatient(String hcPartyId, List<String> secretPatientKeys);

	List<Message> getChildren(String messageId);

	List<Message> getByInvoiceIds(Set<String> invoiceIds);

	List<Message> getByTransportGuids(String hcPartyId, Collection<String> transportGuids);

	List<Message> getByExternalRefs(String hcPartyId, HashSet<String> strings);

	List<Message> listConflicts();

	List<List<Message>> getChildren(List<String> parentIds);
}

File: src/main/kotlin/org/taktik/icure/asyncdao/MessageDAO.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.Message

interface MessageDAO : GenericDAO<Message> {
    fun listMessagesByFromAddressAndActor(partyId: String, fromAddress: String, actorKeys: List<String>?): Flow<Message>
    fun listMessagesByToAddressAndActor(partyId: String, toAddress: String, actorKeys: List<String>?): Flow<Message>
    fun listMessagesByTransportGuidAndActor(partyId: String, transportGuid: String, actorKeys: List<String>?): Flow<Message>

    fun listMessagesByFromAddress(partyId: String, fromAddress: String, paginationOffset: PaginationOffset<List<*>>): Flow<ViewQueryResultEvent>

    fun findMessagesByToAddress(partyId: String, toAddress: String, paginationOffset: PaginationOffset<List<Any>>, reverse: Boolean?): Flow<ViewQueryResultEvent>

    fun findMessagesByHcParty(partyId: String, paginationOffset: PaginationOffset<List<*>>): Flow<ViewQueryResultEvent>

    fun findMessagesByTransportGuid(partyId: String, transportGuid: String?, paginationOffset: PaginationOffset<List<*>>): Flow<ViewQueryResultEvent>

    fun findMessagesByTransportGuidReceived(partyId: String, transportGuid: String?, paginationOffset: PaginationOffset<List<*>>): Flow<ViewQueryResultEvent>

    fun findMessagesByTransportGuidAndSentDate(partyId: String, transportGuid: String, fromDate: Long, toDate: Long, paginationOffset: PaginationOffset<List<Any>>): Flow<ViewQueryResultEvent>

    fun listMessagesByHcPartyAndPatient(hcPartyId: String, secretPatientKeys: List<String>): Flow<Message>
    fun getChildren(messageId: String): Flow<Message>
    fun listMessagesByInvoiceIds(invoiceIds: Set<String>): Flow<Message>
    fun getMessagesByTransportGuids(hcPartyId: String, transportGuids: Collection<String>): Flow<Message>
    fun getMessagesByExternalRefs(hcPartyId: String, externalRefs: Set<String>): Flow<Message>
    fun listConflicts(): Flow<Message>
    fun getMessagesChildren(parentIds: List<String>): Flow<List<Message>>
}
